### PR TITLE
ENH: Add orientation param for text_extraction (# 1071)

### PR DESCRIPTION
**Pull request by pubpub-zz**
add new capability to filter text extraction on orientation

Deprecations: PageObject.extract_text no longer uses the `Tj_sep` and `TJ_sep` parameters.

cf #1071

### CODE REVIEW

File: PyPDF2/_page.py [hunk 1]
```diff
@@ -1106,6 +1106,7 @@ def _extract_text(
         self,
         obj: Any,
         pdf: Any,
+        orientations: Tuple[int] = (0, 90, 270, 360),
```
Comment (MartinThoma) [inline PyPDF2/_page.py:1109]: That default does not match the type signature. For that default, the correct type would be: `Tuple[int, int, int, int]`. If you want a tuple with arbitray arguments, you need `Tuple[int, ...]` (yes, really with the dots :-) )

Alternatively, I guess you want to allow only those 4 numbers. Then `Literal[0, 90, 270, 360]` would be the correct type.

Comment (MartinThoma) [reply PyPDF2/_page.py]: By the way, that can also be combined, e.g.:

`Tuple[Literal[0, 90, 270, 360], ...]`

would mean you have a tuple of arbitrary many integers, but every integer has to be one of 0 / 90 / 270 / 360

Comment (pubpub-zz) [reply PyPDF2/_page.py]: not exactly : What I'm proposing is a variable length tuple or an int):
page.extract_text(0) => extract all text strings oriented up
page.extract_text((0,)) => extract all text strings oriented up (synonym)
page.extract_text((0,180)) => extract all text strings oriented up or down

File: PyPDF2/_page.py [hunk 2]
```diff
@@ -1433,9 +1439,12 @@ def process_operation(operator: bytes, operands: List) -> None:
         return output
 
     def extract_text(
-        self, Tj_sep: str = "", TJ_sep: str = "", space_width: float = 200.0
```
Comment (MartinThoma) [inline PyPDF2/_page.py:1436]: It's pretty unfortunate, but please leave those two parameters in there; maybe with the default `None`. Then add a deprecation notice if they don't have the default value.

It is ok that they will no longer do anything, but I don't want people to have exceptions (I'll add a deprecation note in the commit + in the CHANGELOG)

The new parameter orientations should be at the end if anybody uses positional arguments.

File: PyPDF2/_page.py [hunk 3]
```diff
@@ -1106,6 +1106,7 @@ def _extract_text(
         self,
         obj: Any,
         pdf: Any,
+        orientations: Tuple[int, ...] = (0, 90, 270, 360),
```
Comment (MartinThoma) [inline PyPDF2/_page.py:1109]: Could you please add a it in the docstring with an explanation what it does?

File: PyPDF2/_page.py (hunk 2 above)
Comment (pubpub-zz) [reply PyPDF2/_page.py:1436]: Would it be ok to put them at the end considering they will be used only as named parameters?

Comment (MartinThoma) [reply PyPDF2/_page.py:1436]: How do you know that nobody used it as positional parameters?

Comment (pubpub-zz) [reply PyPDF2/_page.py:1436]: What about this implementation:
```
    def extract_text(
        self,
        *args,
        Tj_sep: str = None, TJ_sep: str = None,
        orientations: Union[int, Tuple[int, ...]] = (0, 90, 270, 360),
        space_width: float = 200.0,
    ) -> str:
        if len(args)>=1:
            if isinstance(args[0],(str,list)):
                warnings.warn("parameters Tj_Sep, TJ_sep depreciated, and will be removed in PyPDF2 3.0.0.",DeprecationWarning)
                Tj_sep =args[0]
                if len(args)>=2 and isinstance(args[1],str):
                    TJ_sep =args[1]
                if len(args)>=3 and isinstance(args[2],(list,int)):
                    orientations = args[2]
            elif isinstance(args[0],(list,int)):
                orientations = args[0]
        if isinstance(orientations, int):
            orientations = (orientations,)
        return self._extract_text(
            self, self.pdf, orientations, space_width, PG.CONTENTS
        )
```

We will have both new and old version 😊

File: PyPDF2/_page.py (hunk 3 above)
Comment (pubpub-zz) [reply PyPDF2/_page.py]: updated with proposed design

**Comment by MartinThoma** [comment]
Very nice!

It looks good to me - I will merge it tomorrow if the text extraction benchmark looks fine as well. So it should get into the next release :-)

**Comment by MartinThoma** [comment]
For `2201.00029` the score increased from 96.7% to 97.7%, the rest stayed the same :+1: 

Interestingly, it seems to have killed a lot of newlines:

image

I think I need to design a new benchmark which captures how well newlines are captured. At the moment, this is completely ignored for calculating the score.

**Comment by MartinThoma** [comment]
However, getting the spaces in / between words right is way more important. And there was the improvement :+1:

**Pull request merged**